[MRG+1] Astype fix

TomDLT opened the pull request:
I extended `check_array` to add support for `accept_sparse=True`
and solve the 3 errors by changing `astype` into `check_array` when sparse matrix are possible

I will check the other uses of `astype` tomorrow.

TODO:
- [x] fix the test error
- [x] finish changing astype to check_array, when we may have a sparse matrix
- [x] clean some remaining usage of warn_if_not_float to use check_array instead

agramfort commented:
thanks @TomDLT 

Travis is not happy. You have failing tests.

TomDLT commented:
How should we change the 'dtype' in the case of sparse matrix?

-'astype' fails in python 3 (cf #4585)
-#4585 suggestion is to use 'check_array', but what do we use 'within check_array'?

TomDLT commented:
OK maybe I will start testing before pushing

ogrisel commented:
Also it would be great to rebase this branch on top of current master while removing the `upstream/master` merge commit by @jrings.

agramfort commented:
> OK maybe I will start testing before pushing
> 

Sounds like a plan ;)

amueller commented:
meh, you can also just check travis after pushing ;) but that utually takes longer.

TomDLT committed:
ENH improve check_array

ENH improve check_array to warn on dtype conversions

ENH make check_array accept several dtypes

ENH change validation with improved check_array

ENH change astype to avoid copy if possible

ENH remove warn_if_not_float

TomDLT commented:
`check_array` is now enhanced to accept :
-`accept_sparse=True` to accept all sparse matrix formats
-`dtype` with an array of accepted types
-`warn_on_dtype=True` to warn if the dtype as been changed

TODO:
-fix the error
-finish changing `astype`to `check_array`, when we may have a sparse matrix
-adapt `check_X_y` for mutiple dtypes
-clean some `warn_if_not_float` using `check_array`

I will look at this monday

ogrisel commented:
To fix the test that fail on travis, you might want to setup a conda env with the same versions for numpy and scipy.

```
conda create -n oldnumpy python=2.6 numpy=1.6.2 scipy=0.11.0
source activate oldnumpy
make clean test
```

ogrisel commented:
Good, catch travis is no green. I re-wrapped and moved your todo list to put it in the description of the PR. Once all the boxes are ticked please edit the title of the PR to replace `[WIP]` (work in progress) to `[MRG]` (ready for merge reviews).

TomDLT commented:
In `k_means.py`, do we want to change ˋ_check_test_dataˋ to use the enhanced version of `check_array`?

It will change a ˋRuntimeWarningˋ into a ˋDataConversionWarningˋ

amueller commented:
I'm not sure how great an idea of ``warn_if_no_float`` was (I think it was mine). We should probably just convert to float.

ogrisel commented:
> I'm not sure how great an idea of warn_if_no_float was (I think it was mine). We should probably just convert to float.

This is what we are doing in this PR when you pass `check_array(X, dtype=[np.float64, np.float32, np.float16])`.

ogrisel commented:
>  In k_means.py, do we want to change ˋ_check_test_dataˋ to use the enhanced version of check_array? It will change a ˋRuntimeWarningˋ into a ˋDataConversionWarningˋ

+1 for using `check_array(X, dtype=FLOAT_DTYPES)` in that case as well. I don't thing that changing the type of warning is such a big deal. Increasing the consistency of the code base is good.

ogrisel commented on sklearn/datasets/twenty_newsgroups.py:
```diff
@@ -345,8 +346,8 @@ def fetch_20newsgroups_vectorized(subset="train", remove=(), data_home=None):
-    X_train = X_train.astype(np.float64)
-    X_test = X_test.astype(np.float64)
+    X_train = astype(X_train, np.float64, copy=False)
+    X_test = astype(X_test, np.float64, copy=False)
```
There is no need to set `copy=False` for this use case as the original data type is `int16` for sure as explained in the previous commit. There is no way to avoid the malloc / dtype conversion in that case. We can keep:

```python
    X_train = X_train.astype(np.float64)
    X_test = X_test.astype(np.float64)
```

ogrisel commented on sklearn/ensemble/gradient_boosting.py:
```diff
@@ -1180,7 +1180,7 @@ def feature_importances_(self):
         self.n_classes_ = 1
         if y.dtype.kind == 'O':
-            y = y.astype(np.float64)
+            y = astype(y, np.float64, copy=False)
```
When `y.dtype.kind == 'O'`, it means that y is an array of Python `object`s. There no need to set `copy=False` as there is no way to avoid the malloc / dtype conversion in that case. Please revert this change.

ogrisel commented on sklearn/manifold/locally_linear.py:
```diff
@@ -43,9 +44,9 @@ def barycenter_weights(X, Z, reg=1e-3):
+        X = astype(X, np.float, copy=False)
     if Z.dtype.kind == 'i':
-        Z = Z.astype(np.float)
+        Z = astype(Z, np.float, copy=False)
```
I think the first two lines:

```python
     X = np.asarray(X)
     Z = np.asarray(Z)
```

should directly be replaced by:

```
    X = check_array(X, dtype=np.float64)
    Z = check_array(Z, dtype=np.float64, allow_nd=True)
```

amueller replied:
+1

amueller replied:
We could also pass through other float types, though. ``dtype=[np.float64, np.float32]``?

ogrisel commented on sklearn/manifold/spectral_embedding_.py:
```diff
@@ -263,7 +264,9 @@ def spectral_embedding(adjacency, n_components=8, eigen_solver=None,
-        laplacian = laplacian.astype(np.float)  # lobpcg needs native floats
+        # lobpcg needs native floats
+        laplacian = check_array(laplacian, dtype=np.float, 
+                                accept_sparse=True, copy=False)
```
We should use `np.float64` to make it explicit that that we want platform independent behavior with double precision floats.

ogrisel commented on sklearn/manifold/spectral_embedding_.py:
```diff
@@ -276,7 +279,9 @@ def spectral_embedding(adjacency, n_components=8, eigen_solver=None,
-        laplacian = laplacian.astype(np.float)  # lobpcg needs native floats
+        # lobpcg needs native floats
+        laplacian = check_array(laplacian, dtype=np.float, 
+                                accept_sparse=True, copy=True)
```
Why `copy=True` here?

amueller replied:
the comment says lobpcg needs native float.... I'm not sure where that comes from, though...

ogrisel commented on sklearn/manifold/spectral_embedding_.py:
```diff
@@ -287,7 +292,8 @@ def spectral_embedding(adjacency, n_components=8, eigen_solver=None,
             # lobpcg needs native floats
-            laplacian = laplacian.astype(np.float)
+            laplacian = check_array(laplacian, dtype=np.float, 
+                                    accept_sparse=True, copy=False)
```
`copy=False` is the default. No need to mention it. Also `np.float64`.

ogrisel commented on sklearn/naive_bayes.py:
```diff
@@ -473,7 +474,7 @@ def partial_fit(self, X, y, classes=None, sample_weight=None):
 
         # convert to float to support sample weight consistently
-        Y = Y.astype(np.float64)
+        Y = astype(Y, np.float64, copy=False)
```
`label_binarize` always returns `np.int64` values for `Y`. Therefore `Y = astype(Y, np.float64, copy=False)` will never be able to avoid the data conversion and extra malloc. To avoid this we should rather refactor  `label_binarize` to accept an extra `dtype` argument but this should better be done in a dedicated PR.

ogrisel commented on sklearn/naive_bayes.py:
```diff
@@ -522,7 +523,7 @@ def fit(self, X, y, sample_weight=None):
         # convert to float to support sample weight consistently;
         # this means we also don't have to cast X to floating point
-        Y = Y.astype(np.float64)
+        Y = astype(Y, np.float64, copy=False)
```
Same remark as https://github.com/scikit-learn/scikit-learn/pull/4645/files#r29588000 but for `LabelBinarizer()` instead of `label_binarize`.

ogrisel replied:
Please revert that change and maybe improve the inline comment to mention that as of now `LabelBinarizer`'s `fit_transform` method returns arrays with `dtype=np.int64`

ogrisel commented on sklearn/utils/random.py:
```diff
@@ -238,7 +238,7 @@ def random_choice_csc(n_samples, classes, class_probability=None,
             raise ValueError("class dtype %s is not supported" %
                              classes[j].dtype)
-        classes[j] = classes[j].astype(int)
+        classes[j] = astype(classes[j], int, copy=False)
```
by default this will return an array with `dtype=np.int64` what ever the `dtype` of `classes[j]`. It's better to make it explicit that we want platform independent behavior with:

```python
        classes[j] = astype(classes[j], np.int64, copy=False)
```

ogrisel commented on sklearn/utils/validation.py:
```diff
@@ -444,7 +492,7 @@ def check_X_y(X, y, accept_sparse=None, dtype="numeric", order=None, copy=False,
         _assert_all_finite(y)
     if y_numeric and y.dtype.kind == 'O':
-        y = y.astype(np.float64)
+        y = astype(y, np.float64, copy=False)
```
This specific change can also be reverted as here `y` is guaranteed to be an array of objects so `copy=False` will never bring anything.

TomDLT commented:
Is there a clean way to revert some old commits without a total rebase?

amueller commented:
"Total rebase"
![total rebase](http://images.amcnetworks.com/ifc.com/wp-content/uploads/2012/08/080312-total-recall.jpg)
At the end, squashing would be nice anyhow.
You can do rebase -i and just delete commits you don't like, too.

TomDLT commented:
And now it's clean

amueller commented:
Thanks :)

ogrisel commented:
LGTM, as @amueller I think you can squash all the commits into a single commit.

ogrisel commented on sklearn/utils/validation.py:
```diff
@@ -232,25 +234,27 @@ def _ensure_sparse_format(spmatrix, accept_sparse, dtype, copy,
+
+    if dtype != spmatrix.dtype:
+        # convert dtype
+        spmatrix = spmatrix.astype(dtype)
```
Actually we need a new boolean flag

```python
    converted_dtype = True
```

ogrisel commented on sklearn/utils/validation.py:
```diff
@@ -232,25 +234,27 @@ def _ensure_sparse_format(spmatrix, accept_sparse, dtype, copy,
+    if dtype != spmatrix.dtype:
+        # convert dtype
+        spmatrix = spmatrix.astype(dtype)
+    elif copy and not changed_format:
```
Then use that flag here to skip the copy if the dtype has been converted:

```python
    elif copy and not changed_format and not converted_dtype:
```

ogrisel replied:
or maybe not actually, because when we ask for a copy we might expect the other parts of the scipy sparse dastructure to be copied as well, not just the array that store the dtype.

ogrisel commented on sklearn/utils/graph.py:
```diff
@@ -113,7 +114,8 @@ def graph_laplacian(csgraph, normed=False, return_diag=False):
                    or np.issubdtype(csgraph.dtype, np.uint)):
-        csgraph = csgraph.astype(np.float)
+        csgraph = check_array(csgraph, dtype=np.float64,
+                              accept_sparse=True, copy=False)
```
`copy=False` is not required as this is the default behavior of `check_array`.

amueller commented on sklearn/manifold/spectral_embedding_.py:
```diff
@@ -263,7 +263,9 @@ def spectral_embedding(adjacency, n_components=8, eigen_solver=None,
         if not sparse.issparse(laplacian):
             warnings.warn("AMG works better for sparse matrices")
-        laplacian = laplacian.astype(np.float)  # lobpcg needs native floats
+        # lobpcg needs native floats
```
so is the comment just not true? I feel this is weird.

amueller replied:
Maybe @GaelVaroquaux knows? ba0c787a48fb3d4ef98a747e7566134c99d2c61a

ogrisel replied:
I think in almost platforms we support `.astype(np.float)` is actually doing the same as `.astype(np.float64)`. It's just that I prefer not having any ambiguity when reading the code. 

I think the comment should be "lobpcg needs double precision floats".

ogrisel replied:
I just tested the following under Python 2.7 32 bit and Python 3.4 64 bit under windows:

```
Python 2.7.9 (default, Dec 10 2014, 12:24:55) [MSC v.1500 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> np.arange(3).astype(np.float).dtype
dtype('float64')
```

```
Python 3.4.3 (v3.4.3:9b73f1c3e601, Feb 24 2015, 22:44:40) [MSC v.1600 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> np.arange(3).astype(np.float).dtype
dtype('float64')
```

amueller commented on sklearn/manifold/spectral_embedding_.py:
```diff
@@ -276,7 +278,9 @@ def spectral_embedding(adjacency, n_components=8, eigen_solver=None,
 
     elif eigen_solver == "lobpcg":
-        laplacian = laplacian.astype(np.float)  # lobpcg needs native floats
+        # lobpcg needs native floats
```
why is this comment here again?

amueller commented on sklearn/manifold/spectral_embedding_.py:
```diff
@@ -287,7 +291,8 @@ def spectral_embedding(adjacency, n_components=8, eigen_solver=None,
             embedding = diffusion_map.T[:n_components] * dd
         else:
             # lobpcg needs native floats
-            laplacian = laplacian.astype(np.float)
```
and here? that is odd...

ogrisel replied:
We can actually remove that check as it's already done before entering the if / else blocks.

amueller commented:
can we remove the ``warn_if_not_float`` helper?

amueller commented on sklearn/utils/validation.py:
```diff
@@ -232,25 +234,27 @@ def _ensure_sparse_format(spmatrix, accept_sparse, dtype, copy,
         Matrix that is ensured to have an allowed type.
     """
-    if accept_sparse is None:
+    if accept_sparse in [None, False]:
```
Why did you add ``False`` as an option?

TomDLT replied:
As I added the possibility to have `accept_sparse=True` so as to accept any sparse format, a user may be tempted to change a `True` into a `False` instead of a `None`

amueller replied:
Fair enough.

ogrisel replied:
Alternatively we could have just added support `accept_sparse='any'`. But True / False are fine too in my opinion.

TomDLT commented:
Indeed, as far as i know, `warn_if_not_float` is not used anymore, do I remove it?

amueller commented:
Run a ``git grep``. If it is not used (and I think it shouldn't be), please remove it.

amueller commented on sklearn/utils/validation.py:
```diff
@@ -322,20 +336,34 @@ def check_array(array, accept_sparse=None, dtype="numeric", order=None,
+        # not a data type (e.g. a column named dtype in a pandas DataFrame)
+        dtype_orig = None
+
+    if dtype_numeric:
```
I'm not 100% sure if we want to keep this behavior. In what cases do we want to pass through integer data? Or bytes? But maybe that is for another PR? I'm not sure....
We could also just set dtype = NUMERIC_DTYPES (which would be a long list) to remove this branch at least.

ogrisel replied:
I also thought about that option but I think it's fine to leave it this way for now.

amueller commented:
This is awesome, thanks!
I am a bit uncertain about the conversion in spectral_embedding, otherwise this LGTM.

I'm not sure how good the "numeric" option was, and maybe we should remove it now, and use explicit dtypes instead? But I think we can merge this before that refactoring.

TomDLT commented:
Do you want me to squash all commits in only one?

ogrisel commented:
Maybe we should deprecate `warn_if_not_float` instead of removing it. I know those utilities are not supposed to be part of the public API but... Any opinion @amueller ?

ogrisel commented:
> I'm not sure how good the "numeric" option was, and maybe we should remove it now, and use explicit dtypes instead? But I think we can merge this before that refactoring.

+1 for thinking about this point again once this PR has been merged.

ogrisel commented:
> Do you want me to squash all commits in only one?

I think so yes.

amueller commented:
I think removing ``warn_if_not_float`` is fine. And we should really split public / private utils :-/

amueller commented:
+1 on squashing everything and +1 on merge.

amueller commented:
See #4685 for tracking the possible removal of "numeric".

TomDLT commented:
Squashed

TomDLT commented:
Just curiosity, to deprecate `warn_if_not_float`, how can I do a relative import of `deprecated` ?
(inside `utils/validation.py`, import from `utils/__init__.py`)

amueller commented:
from . import deprecated maybe?

TomDLT commented:
It did not work for me

amueller commented:
@ogrisel merge?

ogrisel commented:
Fine with the removal of `warn_if_not_float` then. +1 for merge.

Can you please rebase on top of the current master once again? There is a small conflict in `sklearn/preprocessing/data.py` (easy to resolve, just keep the `min_max_axis` import). All tests pass on my box but I would like travis to run once again on this PR before merging.

TomDLT commented:
Rebased on top of master

amueller commented:
Thanks :)

amueller commented:
This is really great btw :beers:

TomDLT commented:
Thanks :)

agramfort commented:
:beers: !

ogrisel commented:
I forgot to drink a beer on that PR: :beers:
